messaging/src/main/java/org/axonframework/messaging/LambdaBasedMessageTypeResolver.java

```diff
@@ -0,0 +1,168 @@
+
+    /**
+     * Private constructor used by the builder to create the resolver with the configured resolvers and default
+     * behavior.
+     *
+     * @param resolvers       The map of payload types to their specific resolvers
+     * @param defaultResolver The default resolver to use when no specific resolver is found, or null to throw an
+     *                        exception
+     */
+    private LambdaBasedMessageTypeResolver(Map<Class<?>, MessageTypeResolver> resolvers,
+                                           MessageTypeResolver defaultResolver) {
+        this.resolvers = resolvers;
+        this.defaultResolver = defaultResolver;
+    }
+
+    /**
+     * Starts building a new resolver and registers the first payload type with a custom resolver.
+     *
+     * @param payloadType The class to register the resolver for
+     * @param resolver    The resolver that resolves the payload type class to a {@link MessageType}
+     * @return The type resolver phase for further configuration
+     */
+    public static TypeResolverPhase on(Class<?> payloadType, MessageTypeResolver resolver) {
+        Map<Class<?>, MessageTypeResolver> initialResolvers = new HashMap<>();
+        initialResolvers.put(payloadType, resolver);
+        return new InternalTypeResolverPhase(initialResolvers, null);
+    }
+
+    /**
+     * Starts building a new resolver and registers the first payload type with a fixed message type.
+     *
+     * @param payloadType The class to register the fixed message type for
+     * @param messageType The message type to use for the given payload type
+     * @return The type resolver phase for further configuration
+     */
+    public static TypeResolverPhase on(Class<?> payloadType, MessageType messageType) {
+        return on(payloadType, payloadType1 -> messageType);
+    }
+
+    @Override
+    public MessageType resolve(Class<?> payloadType) {
+        var resolver = resolvers.get(payloadType);
+        if (resolver != null) {
+            return resolver.resolve(payloadType);
+        }
+        if (defaultResolver == null) {
+            throw new IllegalArgumentException("No resolver found for payload type [" + payloadType.getName() + "]");
+        }
+        return defaultResolver.resolve(payloadType);
+    }
+
+    /**
+     * Interface representing the phase where type resolvers are registered.
+     */
+    public interface TypeResolverPhase {
+
+        /**
+         * Registers a resolver that returns a {@link MessageType} for the given payload type.
+         *
+         * @param payloadType The class to register the resolver for
+         * @param resolver    The resolver that produces a {@link MessageType} for the given payload type
+         * @return The current phase for further configuration
+         * @throws IllegalArgumentException if a resolver is already registered for the given payload type
+         */
+        TypeResolverPhase on(Class<?> payloadType, MessageTypeResolver resolver);
```

Comment:
I am wondering who's going to use this. Can you think of a scenario where somebody would define a `MessageTypeResolver` for a `payloadType`? Wouldn't this just be a mapping from `payloadType` to `MessageType`, and that's it?

Reply:
I just thought it's more flexbile. Let's assume for one of your messages you'd like to use the type name, you can just pass `ClassBasedMessageTypeResolver` to it. But... we don't consider it as a best practice if you base your MessageType on class (it's why we introduced it, to make it decoupled), so... I think I can remove it and left just the mapping. What do you think? Is it obsolete / or even suggesting wrong idea :D

Reply:
I think I will remove it, but maybe I will create also more sophisticated API, where you don't need to repeat the same namespace, like:

```
LambdaBasedMessageTypeResolver
.on(Sample1.class, new MessageType("namespace1", "localName1", "version1")
.namespace(
      "namespace1", 
      namespace -> namespace
                            .on(Sample1.class, "localName1", "version1"))
)
.namespace(
      "namespace1", 
      namespace -> namespace
                                  .version("version1")
                                  .on(Sample1.class, "localName1")
)
```
What do you think about that? All are equivalent to others.

Reply:
```
LambdaBasedMessageTypeResolver
.namespace("namespace")
.on(Type.class, "localName", "version")
.on(Type2.class, "localName", "version")
.namespace("namespace1")
.on(Type3.class, "localName", "version")
```

Reply:
I've done it, but now I'm not sure about that. 
Let's assume we have:
```
NamespaceMessageTypeResolver
                .namespace("namespace1")
                .message(String.class, "test.message", "1.0.0")
                .namespace("namespace2")
                .message(String.class, "test.message", "1.0.0")
```
Now I may expect I can do that, but it's not possible - I have one mapping per type here.